src/WebJobs.Script.WebHost/WebScriptHostBuilderExtension.cs

```diff
@@ -64,42 +64,42 @@ public static class WebScriptHostBuilderExtension
-
-                    // Make sure the registered IHostIdProvider is used
-                    IHostIdProvider provider = rootServiceProvider.GetService<IHostIdProvider>();
-                    if (provider != null)
-                    {
-                        services.AddSingleton<IHostIdProvider>(provider);
-                    }
-
-                    // Logging and diagnostics
-                    services.AddSingleton<IMetricsLogger, WebHostMetricsLogger>();
-                    services.AddSingleton<IEventCollectorProvider, FunctionInstanceLogCollectorProvider>();
-
-                    // Hosted services
-                    services.TryAddEnumerable(ServiceDescriptor.Singleton<IHostedService, HttpInitializationService>());
-                    services.TryAddEnumerable(ServiceDescriptor.Singleton<IHostedService, FileMonitoringService>());
-
-                    ConfigureRegisteredBuilders(services, rootServiceProvider);
+                var webHostEnvironment = rootServiceProvider.GetService<IScriptWebHostEnvironment>();
+                var environment = rootServiceProvider.GetService<IEnvironment>();
+                if (FunctionsSyncManager.IsSyncTriggersEnvironment(webHostEnvironment, environment))
+                {
+                    services.AddSingleton<IHostedService, FunctionsSyncService>();
+                }
+
+                services.AddSingleton<HttpRequestQueue>();
+                services.AddSingleton<IHostLifetime, JobHostHostLifetime>();
+                services.AddSingleton<IWebJobsExceptionHandler, WebScriptHostExceptionHandler>();
+                services.AddSingleton<IScriptJobHostEnvironment, WebScriptJobHostEnvironment>();
+
+                services.AddSingleton<DefaultScriptWebHookProvider>();
+                services.TryAddSingleton<IScriptWebHookProvider>(p => p.GetService<DefaultScriptWebHookProvider>());
+                services.TryAddSingleton<IWebHookProvider>(p => p.GetService<DefaultScriptWebHookProvider>());
+                services.TryAddSingleton<IJobHostMiddlewarePipeline, DefaultMiddlewarePipeline>();
+                services.TryAddEnumerable(ServiceDescriptor.Singleton<IJobHostHttpMiddleware, CustomHttpHeadersMiddleware>());
+                services.TryAddEnumerable(ServiceDescriptor.Singleton<IJobHostHttpMiddleware, HstsConfigurationMiddleware>());
+                services.TryAddSingleton<IScaleMetricsRepository, TableStorageScaleMetricsRepository>();
+
+                // Make sure the registered IHostIdProvider is used
+                IHostIdProvider provider = rootServiceProvider.GetService<IHostIdProvider>();
+                if (provider != null)
+                {
+                    services.AddSingleton<IHostIdProvider>(provider);
+                }
+
+                // Logging and diagnostics
+                services.AddSingleton<IMetricsLogger>(a => new WebHostMetricsLoggerDecorator(metricsLogger));
```

Comment:
@fabiocav Is there any other way to inject this without running into circular dependency issues?

Reply:
Why does this have to be a decorator? We should be able to wire up the ordering such that this one "wins". Can you use `TryAddSingleton<IMetricsLogger>` in the ScriptHost setup, for example?

Reply:
That was kind of what I went with initially. 

Something like services.AddSingleton(rootServiceProvider.GetService<IMetricsLogger>());. But the problem with that is, jobHost disposal (during restarts) ends up disposing the rootService's WebHostMetricsLogger thereby breaking metrics altogether.

---
In reply to: [335982314](https://github.com/Azure/azure-functions-host/pull/5084#discussion_r335982314) [](ancestors = 335982314)

Reply:
That shouldn't dispose... if you use a factory, it will (i.e. `AddSingleton(s => ...)`) but if you pass in an existing instance, it shouldn't dispose. That's what we do above for `IHostIdProvider`. We take it from the `rootServiceProvider` and explicitly register it with the ScriptHost.

Can you try that pattern again and see if it works? It'd remove this decorator class, which would be more straightforward.